src/object/srv_ec_aggregate.c

```diff
@@ -1900,16 +1920,23 @@ agg_process_stripe(struct ec_agg_param *agg_param, struct ec_agg_entry *entry)
 		goto out;
 	}
 
-	/* With parity and some newer partial replicas, possibly holes */
-	if (ec_age_with_hole(entry))
-		process_holes = true;
-	else
-		rc = agg_process_partial_stripe(entry);
+	/* With parity and some newer partial replicas, possibly holes.
+	 * Case 1: with valid hole (hole epoch >= parity epoch), can be handled by
+	 *         agg_process_holes().
+	 * Case 2: without valid hole, must with partial replica, should be handled by
+	 *         agg_process_partial_stripe().
+	 */
+	if (ec_age_with_hole(entry)) {
```

Comment:
I'm not sure if I fully followed the changes for the hole processing.

It looks to me we just need to save the highest hole epoch in agg_extent_add(), and modify the ec_age_with_hole() to compare the highest hole epoch with parity epoch to see if we need process holes? (Then all other changes could be avoided?)

Reply:
it is related with hole extent carry over case as we discussed, probably can cleanup those stuff after 2.6. thx